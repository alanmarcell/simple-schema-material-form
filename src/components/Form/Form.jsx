import React from 'react';
import PropTypes from 'prop-types';
import { pathOr, merge, isNil, isEmpty, filter } from 'ramda';
import { compose, withStateHandlers, pure, withState } from 'recompact';
import { FormControl } from 'material-ui/Form';
import { Button } from 'material-ui';
import handleSubmit from './submitHandler';

const hasError = (errors = [], fieldName) => {
  if (isNil(errors)) {
    return false;
  }

  const getError = error => {
    if (error.name === fieldName) {
      return error;
    }

    return null;
  };

  const fieldError = filter(getError, errors);

  if (isEmpty(fieldError)) {
    return false;
  }

  return true;
};

const getChildProps = ({ doc, setDoc, errors }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    error: hasError(errors, fieldName),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

/** Our Simple Form Test Utility */
const SimpleForm = (props) => {
  const { doc, setDoc, errors } = props;
  const setChildProps = getChildProps({ doc, setDoc, errors });
  return (
    <FormControl >
      {React.Children.map(props.children, setChildProps)}
      <Button onClick={() => handleSubmit(props)} >Submit</Button>
    </FormControl>
  );
};

/** This is our base enhancer to handle doc changes */
const enhance = compose(
  pure,
  withState('errors', 'setErrors', []),
  withStateHandlers(
    /** this our initial doc */
    { doc: {} },
    {
      /** this is the doc state handler */
      setDoc: ({ doc }) => value => {
        const merged = merge(doc, value);
        return {
          doc: merged,
        };
      },
    },
  ),
);

const EnhancedSimpleForm = enhance(SimpleForm);

SimpleForm.defaultProps = {
  /** This is our fields */
  // children: null,
};

SimpleForm.propTypes = {
  /** doc defines the form state */
  doc: PropTypes.object.isRequired,

  /** handle the form state */
  setDoc: PropTypes.func.isRequired,

  children: PropTypes.node.isRequired,
};

export { SimpleForm };
export default EnhancedSimpleForm;
