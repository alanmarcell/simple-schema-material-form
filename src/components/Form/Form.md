# Form example

```jsx
<Form name="Hello" >
  <TextInput fieldName="name" />
</Form>
```
