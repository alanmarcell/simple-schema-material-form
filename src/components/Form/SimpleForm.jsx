import React from 'react';
import PropTypes from 'prop-types';
import { omit, pathOr, merge } from 'ramda';
import { compose, withStateHandlers, pure } from 'recompact';
import { Button, TextField } from 'material-ui';

const getChildProps = ({ doc, setDoc }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

export const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

/** Our Simple Form Test Utility */
const SimpleForm = (props) => {
  const { doc, setDoc, onSubmit } = props;
  const setChildProps = getChildProps({ doc, setDoc });
  return (
    <form >
      {React.Children.map(props.children, setChildProps)}
      <Button onClick={() => onSubmit(doc)} >Submit</Button>
    </form>
  );
};

/** This is our base enhancer to handle doc changes */
const enhance = compose(
  pure,
  withStateHandlers(
    /** this our initial doc */
    { doc: {} },
    {
      /** this is the doc state handler */
      setDoc: ({ doc }) => value => {
        const merged = merge(doc, value);
        return {
          doc: merged,
        };
      },
    },
  ),
);

const EnhancedSimpleForm = enhance(SimpleForm);

SimpleForm.defaultProps = {
  /** This is our fields */
  // children: null,
};

SimpleForm.propTypes = {
  /** doc defines the form state */
  doc: PropTypes.object.isRequired,

  /** handle the form state */
  setDoc: PropTypes.func.isRequired,

  children: PropTypes.array.isRequired,
};

export default EnhancedSimpleForm;
