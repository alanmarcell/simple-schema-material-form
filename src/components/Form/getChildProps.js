import React from 'react';
import { pathOr, isNil, isEmpty, filter } from 'ramda';

const getFieldError = fieldName => error => {
  if (error.name === fieldName) {
    return error;
  }
  return null;
};

const hasError = (errors = [], fieldName) => {
  if (isNil(errors)) {
    return false;
  }

  return !isEmpty(filter(getFieldError(fieldName), errors));
};


const getChildProps = ({ doc, setDoc, errors }) => Child => {
  const { fieldName } = Child.props;
  const value = pathOr('', [fieldName], doc);

  const childProps = {
    ...Child.props,
    setDoc,
    doc,
    error: hasError(errors, fieldName),
    value,
    label: fieldName,
  };

  return React.createElement(
    Child.type,
    childProps,
    null,
  );
};

export default getChildProps;
