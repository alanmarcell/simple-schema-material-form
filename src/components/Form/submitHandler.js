import validateDoc from './validateForm';

const handleSubmit = ({
  buildDocToSave = doc => doc,
  buildDocToValidate = doc => doc,
  ...props
}) => {
  const docIsValid = validateDoc(buildDocToValidate(props));
  const {
    onSubmit,
  } = props;

  if (docIsValid) {
    const docToSave = buildDocToSave(props.doc);

    onSubmit(docToSave);
  }
};

export default handleSubmit;
