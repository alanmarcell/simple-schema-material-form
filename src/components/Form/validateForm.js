const validateDoc = ({ schema, doc = {}, setErrors }) => {
  if (!schema) {
    return true;
  }

  const validationContext = schema.newContext();
  const validation = validationContext.validate(doc);

  const docErrors = validationContext.validationErrors();
  setErrors({ ...docErrors, submitAttempt: true });
  return validation;
};

export default validateDoc;
