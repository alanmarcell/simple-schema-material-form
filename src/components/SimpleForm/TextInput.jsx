import React from 'react';
// import PropTypes from 'prop-types';
import { omit, pathOr } from 'ramda';
// import { compose, withStateHandlers, pure } from 'recompact';
import { TextField } from 'material-ui';


const TextInput = props => {
  const { doc, setDoc, fieldName } = props;
  const value = pathOr('', [fieldName], doc);
  const cProps = omit(['fieldName', 'setDoc', 'doc'], props);

  const inputProps = {
    ...cProps,
    onChange: e => setDoc({ [fieldName]: e.target.value }),
    value,
    label: fieldName,
  };
  return <TextField {...inputProps} />;
};

export default TextInput;
