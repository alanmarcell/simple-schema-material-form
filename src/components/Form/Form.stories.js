import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withState } from '@dump247/storybook-state';
import { merge } from 'ramda';
import Typography from 'material-ui/Typography';
import SimpleSchema from 'simpl-schema';

import { SimpleForm } from './Form';
import TextInput from '../TextInput';
import Checkbox from '../Checkbox';
import Radio from '../Radio';
import withTests from '../../stories/withTests';

const setDoc = store => doc => store.set({ doc: merge(store.state.doc, doc) });

const Form = ({ store, children }) => (
  <SimpleForm
    {...store.state}
    setDoc={setDoc(store)}
    onSubmit={action('doc')}
  >
    {children}
  </SimpleForm>
);

const SessionSchema = new SimpleSchema({
  required: {
    type: String,
    min: 6,
  },
});

storiesOf('Form|TextInput', module)
  .addDecorator(withTests('Form'))
  .add('with TextInput Schema Validator', withState({ doc: {} }, (store) => (
    <div style={{ display: 'flex', alignItems: 'space-between', flexDirection: 'column' }}>
      <Typography variant="title" gutterBottom>
        If you hit submit without the required field a error is displayed
      </Typography>
      <Form store={store} schema={SessionSchema} >
        <TextInput fieldName="required" />
        <TextInput fieldName="optional" />
      </Form>
    </div>
  )))
  .add('with TextInput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <TextInput fieldName="test" />
    </Form>
  )))
  .add('with 2 TextInput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <TextInput fieldName="test" />
      <TextInput fieldName="test2" />
    </Form>
  )));


storiesOf('Form|Checkbox', module)
  .addDecorator(withTests('Form'))
  .add('with Checkbox', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
    </Form>
  )))
  .add('with Checkbox and Text Iput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Checkbox selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
      <TextInput fieldName="test2" />
    </Form>
  )));

storiesOf('Form|Radio', module)
  .addDecorator(withTests('Form'))
  .add('with Radio', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Radio selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
    </Form>
  )))
  .add('with Checkbox and Text Iput', withState({ doc: {} }, (store) => (
    <Form store={store}>
      <Radio selectOptions={[{ label: 1, key: 'um' }, { label: 2, key: 'dois' }]} fieldName="test" />
      <TextInput fieldName="test2" />
    </Form>
  )));
