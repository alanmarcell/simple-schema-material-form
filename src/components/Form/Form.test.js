import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { Button } from 'material-ui';
import SimpleSchema from 'simpl-schema';
import Form from './Form';
import TextInput from '../TextInput';

describe('Form', () => {
  it('should renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Form ><TextInput fieldName="test" /></Form>, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  describe('TextInput', () => {
    const SessionSchema = new SimpleSchema({
      // required: {
      //   type: String,
      //   required: true,
      // },
      minRequired: {
        type: String,
        min: 6,
      },
      // optional: {
      //   type: String,
      // },
    });
    describe('Minimun required', () => {
      const onSubmit = jest.fn();
      const FormComponent = (
        <Form schema={SessionSchema} onSubmit={doc => onSubmit(doc)}>
          <TextInput fieldName="minRequired" />
        </Form>);
      const component = shallow(FormComponent);
      const newValue = 'typ.aaa.';
      let { children } = component.props();
      let Children = shallow(children[0][0]);
      Children.simulate('change', {
        target: { value: newValue },
      });

      let RequiredTextInput = component.find({ label: 'minRequired' });

      let button = component.find(Button);
      button.simulate('click');

      RequiredTextInput = component.find({ label: 'minRequired' });

      describe('With error', () => {
        it('should add error prop if required field dont reach min required', () => {
          expect(RequiredTextInput.props().error).toBe(true);
        });

        it.skip('submit should not be invoked if have errors', () => {
          expect(onSubmit).not.toHaveBeenCalled();
        });
      });

      describe('Without error', () => {
        ({ children } = component.props());
        Children = shallow(children[0][0]);

        let ValidTextInput = component.find({ label: 'minRequired' });
        Children.simulate('change', {
          target: { value: 'typing' },
        });

        component.update();
        button = component.find(Button);

        button.simulate('click');

        component.update();
        ValidTextInput = component.find({ label: 'minRequired' });

        it('should remove error prop if required field reach min required', () => {
          expect(ValidTextInput.props().error).toBe(false);
        });

        it('submit should be invoked if have no errors', () => {
          expect(onSubmit).toHaveBeenCalled();
        });
      });
    });


    it('shallow a TextInput', () => {
      const onSubmit = jest.fn();
      const FormComponent = (
        <Form onSubmit={doc => onSubmit(doc)}>
          <TextInput fieldName="test" />
        </Form>);

      const component = shallow(FormComponent);
      const newValue = 'Type Sample';

      const { children } = component.props();

      const Children = shallow(children[0][0]);

      Children.simulate('change', {
        target: { value: newValue },
      });

      component.update();
      const newState = component.state();
      const newChildren = component.props().children[0][0];
      const NewChildren = shallow(newChildren);

      const { doc } = newState.childProps;
      const test2 = newState.childProps.doc.test;

      expect(test2).toBe(newValue);
      expect(NewChildren.props().value).toBe(newValue);
      expect(component).toMatchSnapshot();


      const button = component.find(Button);
      button.simulate('click');

      expect(onSubmit).toHaveBeenCalled();
      expect(onSubmit).toBeCalledWith(doc);
    });


    it('shallow a 2 TextInput', () => {
      const onSubmit = jest.fn();
      const FormComponent = (
        <Form onSubmit={doc => onSubmit(doc)}>
          <TextInput fieldName="test" />
          <TextInput fieldName="test2" />
        </Form>);

      const component = shallow(FormComponent);

      const newValue = 'Type Sample';
      const { children } = component.props();
      const TextInput1 = shallow(children[0][0]);
      const TextInput2 = shallow(children[0][1]);

      TextInput1.simulate('change', {
        target: { value: newValue },
      });
      TextInput2.simulate('change', {
        target: { value: newValue },
      });
      const newState = component.state();
      const { doc } = newState.childProps;

      component.update();

      const newChildren = component.props().children;
      const NewTextInput1 = shallow(newChildren[0][0]);
      const NewTextInput2 = shallow(newChildren[0][1]);

      const button = component.find(Button);
      button.simulate('click');

      const test2 = newState.childProps.doc.test;

      const expectedDoc = {
        test: newValue,
        test2: newValue,
      };

      expect(doc).toEqual(expectedDoc);

      expect(test2).toBe(newValue);
      expect(NewTextInput1.props().value).toBe(newValue);
      expect(NewTextInput2.props().value).toBe(newValue);
      expect(component).toMatchSnapshot();
      expect(onSubmit).toHaveBeenCalled();
      expect(onSubmit).toBeCalledWith(doc);
    });
  });
});
