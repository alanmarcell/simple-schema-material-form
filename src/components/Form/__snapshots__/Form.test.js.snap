// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form TextInput shallow a 2 TextInput 1`] = `
<WithStyles(FormControl)>
  <pure(TextInput)
    doc={
      Object {
        "test": "Type Sample",
        "test2": "Type Sample",
      }
    }
    error={false}
    fieldName="test"
    key=".0"
    label="test"
    setDoc={[Function]}
    value="Type Sample"
  />
  <pure(TextInput)
    doc={
      Object {
        "test": "Type Sample",
        "test2": "Type Sample",
      }
    }
    error={false}
    fieldName="test2"
    key=".1"
    label="test2"
    setDoc={[Function]}
    value="Type Sample"
  />
  <WithStyles(Button)
    onClick={[Function]}
  >
    Submit
  </WithStyles(Button)>
</WithStyles(FormControl)>
`;

exports[`Form TextInput shallow a TextInput 1`] = `
<WithStyles(FormControl)>
  <pure(TextInput)
    doc={
      Object {
        "test": "Type Sample",
      }
    }
    error={false}
    fieldName="test"
    key=".0"
    label="test"
    setDoc={[Function]}
    value="Type Sample"
  />
  <WithStyles(Button)
    onClick={[Function]}
  >
    Submit
  </WithStyles(Button)>
</WithStyles(FormControl)>
`;
